### DISCUSSION TITLE
Where to put test data

### DISCUSSION THREAD

**Discussion by haipinglu** [pykale/pykale#128] [Testing] [answered]
Current pykale package size (compressed):               967KB (1.38MB w/o compression)
Among which the test gait data size (compressed):   275KB (277KB w/o compression)
If adding 776KB of test data, more than half of the pykale package size (compressed) will be test data, which may not be what we want.

There are two options for the test data
- Direct upload: 776k seems fine. The current gait data is 277k. This is <3x. It will be good that this will be reused in other tests in future. For uploaded data, they should be reused in many to worth the fixed presence.
- Host externally and download during tests: this is what Raivo did with the video loader example, line 17-35 in https://github.com/pykale/pykale/blob/main/examples/video_loading/main.py. Actually, this code could be highly reusable and we could consider refactoring it into `kale.loaddata` or `kale.utils` so that reuse will just need to supply the ID and filename. 

Eventually, it should be a balance between repo size and the convenience of coding (any other factor?).

**Comment by bobturneruk** [comment]
I've made an enquiry with a colleague at the library.

**Reply by rosiehigman** [reply]
Hi Bob, I think @haipinglu has outlined the main trade-off. The only thing I would add is if the visibility/longevity of the data is an issue then a data repo would be a better bet. You could deposit it in ORDA and use Figshare's API: https://docs.figshare.com/ to retrieve the data. This would make the data citeable, supports versioning and enables long-term preservation. 

The trade-off being that I think you might have slightly longer load times due to importing the data each time. From a library point of view we are always looking to encourage the preservation of research data, and if the test data will be reused then this is worthwhile.

**Reply by bobturneruk** [reply]
Thanks @rosiehigman!

**Comment by bobturneruk** [comment]
Git annex and DVC are things to consider.

**Comment by bobturneruk** [comment, marked as answer]
Perhaps a seperate repo under the PyKale org just for test data.

**Reply by xianyuanliu** [reply]
If easy to create and maintain one, this will be helpful. Store the data in data repo, download data in test running, and add the data directory to .gitignore. Is it a good way when our test data is not too big yet? We can use other more efficient tools when we have to load big data.

**Reply by bobturneruk** [reply]
It kind of delays having to deal with the problem - we don't accumulate baggage in PyKale and we have a reliable source for the data. Maybe longer term, if datasets get bigger, we need to rethink.

**Reply by haipinglu** [reply]
Great suggestion! Here it is https://github.com/pykale/data @XianyuanLiu

**Reply by haipinglu** [reply]
Individual file size can be up to possibly 80MB (from Cameron, not official) at least. I did not find the official limit. As long as the individual file sizes are not big, it will be fine. Otherwise, triggering Github's Large File Storage (LFS) will incur cost, but no worry they won't charge before we agree, i.e. they will refuse if we run out of free quota  (we have 60MB left).

It is also more accessible than Google Drive (e.g. in China).
